### PR TITLE
Introduce `ResponseConnector` as a superset of `HttpConnector`

### LINKED ISSUE #1085: Deprecate all usages of HttpConnector and HttpURLConnection
[HttpConnector](https://github.com/hub4j/github-api/blob/master/src/main/java/org/kohsuke/github/HttpConnector.java) depends on [HttpURLConnection](https://docs.oracle.com/javase/8/docs/api/java/net/HttpURLConnection.html) an old and very crufty way of working with Http connections.  Some modern clients have even stopped supporting HttpURLConnection. This makes it harder to implement support for different http clients. 

We should remove all dependency on `HttpConnector`.   `GithubClient` will become the new base class to extend in order to support new clients.  Other methods the depend on `HttpURLConnection` should switch to other ways of providing the same information.

### PR DESCRIPTION
**Pull request by nedtwigg**
This PR touches a lot of files, but it's pretty straightforward. The commits tell the story as simply as I can tell it, every commit in the chain passes `mvn test`. It resolves #1085 in a backward-compatible way (along with #1262 and #754).

It introduces a new interface `ResponseConnector`, with just a single method which functions as a delegate to  the `getResponseInfo` method of `GitHubClient` which allows `GitHubClient` to be become final.

https://github.com/hub4j/github-api/blob/7a91e436c6493cc607922c849cc8032284282de8/src/main/java/org/kohsuke/github/GitHubClient.java#L403-L406

`GitHubHttpUrlConnectionClient` now implements `ResponseConnector` rather than `GitHubClient`. This means that all of the old `HttpConnector` stuff can keep working transparently:

https://github.com/hub4j/github-api/blob/7a91e436c6493cc607922c849cc8032284282de8/src/main/java/org/kohsuke/github/GitHubClient.java#L137-L143

In order to make it possible to implement `ResponseConnector`, I had to extract `ResponseInfo` into its own class and make it and some of its members public. Most of the commits in this PR are pure refactors (refactor commits aren't mixed in with new or changed functionality) so you can see exactly which parts had to become public.

The whole point of the PR is to bypass `HttpConnector` and users can now swap the new [`OkHttpResponseConnector`](https://github.com/nedtwigg/github-api/blob/remove-http-connector/src/main/java/org/kohsuke/github/extras/okhttp3/OkHttpResponseConnector.java) instead of ([`OkHttpConnector`](https://github.com/nedtwigg/github-api/blob/remove-http-connector/src/main/java/org/kohsuke/github/extras/okhttp3/OkHttpConnector.java) + [`GitHubHttpUrlConnectionClient`](https://github.com/nedtwigg/github-api/blob/remove-http-connector/src/main/java/org/kohsuke/github/GitHubHttpUrlConnectionClient.java)).

I believe the only breaking change in this PR is to `[Abuse/Rate]LimitHandler`, but it's a transparent change to all users who are using the built-in policies, and an easy change for users who aren't. The new names in the PR aren't great (e.g. `ResponseInfo`, `ResponseConnector`) but I just kept all names to match their name from where they were refactored from. I didn't mark any of the older stuff as deprecated.

One path forward might be to:

1. Merge and release this PR (with whatever feedback you'd like)
2. Let it rack up some mileage with users
3. Deprecate `HttpConnector` and all its usages
4. Remove the deprecations in a breaking change, and possibly rename classes at that point as well

### CODE REVIEW

**Comment by nedtwigg** [comment]
Published in `3.33.2`.

**Comment by bitwiseman** [comment]
Okay, this is embarrassing, but I'm working on basically the same change. I haven't looked at your change in any depth, but the description sounds much the same. I'll get you feedback shortly.

**Comment by nedtwigg** [comment]
Lol! Well feel free to dump this PR and just merge yours :) All I care about is that I can do something like `OkHttpResponseConnector`

**Comment by bitwiseman** [comment]
@nedtwigg 
So, yeah, we basically had the same idea. 
I got around the coverage problem by running part of the regular test pass on okhttp.  
Some naming differences of course.  
I also made the change non-breaking for both API and default behavior.

Would you be willing to review #1290 ? It should look very familiar. 😄

**Comment by nedtwigg** [comment]
Absolutely! Review incoming.

File: src/main/java/org/kohsuke/github/GitHubClient.java [hunk 1]
```diff
@@ -409,12 +401,28 @@ private void logRequest(@Nonnull final GitHubRequest request) {
     }
 
     @Nonnull
-    protected abstract GitHubResponse.ResponseInfo getResponseInfo(GitHubRequest request) throws IOException;
-
-    protected abstract void handleLimitingErrors(@Nonnull GitHubResponse.ResponseInfo responseInfo) throws IOException;
+    protected ResponseInfo getResponseInfo(GitHubRequest request) throws IOException {
+        return connector.getResponseInfo(request, authorizationProvider);
+    }
```
Comment (nedtwigg) [inline src/main/java/org/kohsuke/github/GitHubClient.java:406]: I chose to treat the request as unmodifiable, whereas your `GitHubClient` has `private GitHubRequest prepareRequest(GitHubRequest request) throws IOException {`. In that method you add the auth headers, and also do some accept/encoding stuff that I don't understand very well. Probably my PR is just plain missing those features.

I think this way of handling authorization by passing the provider vs modifying the request is the only substantive difference between the two PRs, and I like your way better.

**Comment by nedtwigg** [comment]
I like yours better :) Closing as dupe, looking forward to the next release with the new features.

**Comment by bitwiseman** [comment]
> I chose to treat the request as unmodifiable, whereas your `GitHubClient` has `private GitHubRequest prepareRequest(GitHubRequest request) throws IOException {`. In that method you add the auth headers, and also do some accept/encoding stuff that I don't understand very well. Probably my PR is just plain missing those features.

> I think this way of handling authorization by passing the provider vs modifying the request is the only substantive difference between the two PRs, and I like your way better.

Requests are immutable, this creates a new request.  I should change the name of the method.

**Comment by nedtwigg** [comment]
Fwiw I think your name `prepare` is good. I'll restate the case for passing `AuthorizationProvider`: `GitHubRequest` could be a type-level guarantee "this is what the client code requested" vs "this might be what the client code requested, or it might be a mutated copy with auth information and content-negotiation appended on". I was treating `GitHubRequest` as type-enforced pure-intent, but I think your approach is more practical.

**Pull request closed without merging**